README.md
# satellite_cnns
Training land cover classifications through google earth engine with CNNs


CHANGELOG.md
# dd worknotes:
GC == Google Cloud
GS == Google Storage

Adjusted for [weather AI notebook](https://github.com/GoogleCloudPlatform/python-docs-samples/tree/main/people-and-planet-ai/weather-forecasting)

## Currently Working on: tensorflow support

## To Do
[ ] Now that we're using NPZ files, test if we need to convert labels to float64 (`as_double` functionality in Benin [data.py](src/benin-data/benin/data.py))\
[ ] Change all float64s to float32s to match better with Pytorch (create_dataset change please)\
[ ] Adjust train.py to support different dataset shapes (num of input and label bands)\
[ ] Add logger in train.py\
[x] Read/test uploaded NPZ files\


## Future improvements
[ ] Support for TensorFlow framework\
[ ] Parallel processing for reading data files (read_data.py)\
[ ] Weather sample augments it's examples, good to implement in future training
[ ] Save created model directly to VertexAI
  - Ref "Training did not produce a Managed Model returning None. Training Pipeline projects/978289642310/locations/us-central1/trainingPipelines/5677616430787330048 is not configured to upload a Model. Create the Training Pipeline with model_serving_container_image_uri and model_display_name passed in. Ensure that your training script saves to model to os.environ['AIP_MODEL_DIR']."

## 7/20:
- Tried creating Docker image locally in Colab
  - Docker too convoluted to use in cells
  - Image too large to upload each job
  - A waste to spend time and resources re-building with same base image but different data and model packages
- Tried using pre-built personal Docker Hub image
  - Failed to access -- still might be a solution, but can't get it to work currently. Could use Google's Artifact registry but costs money.
- Tried just adding tensorflow in requirements.txt so it's installed before the job runs. 
  - For some reason getting worker error "A n1-standard-1 VM instance is currently unavailable in the us-central1-a zone. Alternatively, you can try your request again with a different VM hardware configuration or at a later time. For more information, see the troubleshooting documentation." Check [page](https://cloud.google.com/compute/docs/resource-error).
  - Error could be that tensorflow is trying to use resources that isn't in the GC quota or it needs resources that aren't given to it.

## 7/18:
- Started Docker SDK

## 7/16:
- Started implementing tensorflow framework to use previous

## 7/14:
- Worked on predictions
  - Notebook does patch by batch, more difficult to batch patches for land just in Benin geometry


## 7/6:
- Continued work on U-Net implementation and train.py
- Produced a PyTorch model with train.py locally (model training effectiveness untested)
  - finished proof of concept implementation of benin-model package
- Ran traning on VertexAI
  - took npz files from GS bucket 
- Added more helpful to model
- Added basic notebook annotaiton update for training demonstration

## 7/5:
- Started Pytorch U-Net implementation and package (benin-model) to use for training (train.py)
  - Moving away from weather forcasting implementation (more object oriented) to avoid using Hugging Face Trainer. Instead implementing U-Net (https://github.com/milesial/Pytorch-UNet)

## 7/4:
- Finished and tested torch.utils.data.Dataset custom subclass (DatasetFromPath)
- Tested workflow up to use of DatasetFromPath and splitting the dataset
 - Modified notebook to reflect work

## 7/3:
- Created library [read_data.py](read_data.py) for reading the data in NPZ format
- Began translation of weather forcasting reading NPZ, read_dataset to avoid use of Hugging Face Datasets

## 7/2:
- Updated package package file names (benin_data.py -> data.py, etc)
- Changed `SCALE` in [benin/data.py](src/benin-data/benin/data.py) to split to `SAMPLE_SCALE` and `PATCH_SCALE`
  - Stratified sampling at too low of a scale is very memory intenisve and that kind of precision isn't needed. We can increase the strat sampling scale as long as we make sure that the patches retireved from EE is scaled... to the scale (ex: if patch scale is 10, sample scale can be 10,100,1000, etc)
- Cleaned up [create_dataset.py](create_dataset.py)
  - Specified what should be customized to change defaults
- Tested inputs and labels patch using new `show_patches` function (to be added to a visualize.py file)
- Added and tested new pipeline in [create_dataset.py](create_dataset.py)
  - Uses compressed NumPy files instead of TFrecord. This fixes the missing tensorflow module using Dataflow (#4) because it isn't required.
  - WORKING DISTRIBUTED RUNNER
- Changed [benin_apache_pipeline.ipynb](benin_apache_pipeline.ipynb) to reflect current project state.

## 7/1:
- Changed setup to clone SPIRES repo
- Refactored to use repo ^ packages (Added src/)
- Tested DataFlow with [create_dataset.py](create_dataset.py) script
  - Fixed [#4](/../../issues/4) with built and attached data package
- TODO: General refactoring, probably need stricter version requirement checks
- **Current issue**: `serialize_tensorflow` seems to be failing in DataFlow pipeline. See [failed jobs](https://console.cloud.google.com/dataflow/jobs?project=ls-test-3-24&authuser=0) in GC console.
  - Possible solution: packaging things better
  - TODO: test pipeline output locally, though no serialization errors when running local pipeline.  
- Another issue, though less important: Max workers for pipeline is 1 even though it's explicitly set as higher
  - Possible solution: idk; maybe GC setting

## 6/30:
- Fixed visualization to reflect `get_*_image` functionality
- Removed weather-forcasting version of pipeline, replaced with original working pipeline (NumPy -> tfrecord)


## 6/29
Current issue in running the pipeline locally ("EEException: Invalid number of coordinates: 1 [while running '[165]: 📑 Get example']"). Currently stack tracing to try to discover issue.
- Added `SCALE`

src/benin-data/benin/data.py
# This file includes work and structure covered by the following copyright and permission notices:
#
#  Copyright 2023 Google LLC
#  Licensed under the Apache License, Version 2.0 (the "License");
#
# ==============================================================================================================
#
# Data retrieval code for Benin for SPIRES Lab Projects.
#
#
# Modify this file to change what create_dataset.py creates the testing/training datset from.
#
# Project patterns based on weather-forcasting sample:
# https://github.com/GoogleCloudPlatform/python-docs-samples/tree/main/people-and-planet-ai/weather-forecasting
#
# ______________________________________________________________________________________________________________



# For posponed evaluations
#from __future__ import annotations

import io
import ee
from google.api_core import exceptions, retry
import google.auth
import requests
from typing import Iterable

import numpy as np
from numpy.lib.recfunctions import structured_to_unstructured


# Authenticate and initialize Earth Engine with the default credentials.
credentials, project = google.auth.default(
    scopes=[
        "https://www.googleapis.com/auth/cloud-platform",
        "https://www.googleapis.com/auth/earthengine",
    ]
)

# Use the Earth Engine High Volume endpoint.
#   https://developers.google.com/earth-engine/cloud/highvolume
ee.Initialize(
    credentials.with_quota_project(None),
    project=project,
    opt_url="https://earthengine-highvolume.googleapis.com",
)


# Default globals
PATCH_SCALE = 10 
SAMPLE_SCALE = 1000
assert(SAMPLE_SCALE % PATCH_SCALE == 0) # PATCH_SCALE should be multiple of SAMPLE_SCALE

#OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO
# The following code should be edited depending on the project.
# 
# To be a funcitonal file for creating a dataset, define
#   - get_inputs_image()-> ee.Image
#   - get_labels_image()-> ee.Image
#   - sample_points(seed: int, points_per_class: int) -> Iterable[tuple[float, float]]
#
#OOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO

def mask_clouds_landsat(image: ee.Image) -> ee.Image:
    """
    Function credit unknown @LS

    """
    # Bits 3 and 5 are cloud shadow and cloud, respectively
    cloudShadowBitMask = (1 << 3) # 1000 in base 2
    cloudsBitMask = (1 << 5) # 100000 in base 2

    # Get the pixel QA band
    qa = image.select('pixel_qa')

    #Both flags should be set to zero, indicating clear conditions
    mask = qa.bitwiseAnd(cloudShadowBitMask).eq(0).bitwiseAnd(qa.bitwiseAnd(cloudsBitMask).eq(0))

    # Mask image with clouds and shadows
    return image.updateMask(mask)


def get_inputs_image() -> ee.Image:
    """ Retrieves the input image for Benin data

    Uses Landsat7

    Args:
      as_double (bool): Whether to cast the image 
          to a double. This is done to match the 
          dtype of the input image

    Returns:
      ee.Image: inputs with bands 'R', 'G', 'B', 'NDVI'
        - has benin geometry with 10000 buffer

    """

    # Specify inputs (Landsat bands) to the model and the response variable.
    opticalBands = ['B3','B2','B1'] #RGB
    thermalBands = ['B4','B3'] #NIR

    # Grab the Benin feature (shape of country)
    benin_shape = ee.FeatureCollection("USDOS/LSIB_SIMPLE/2017").filter(ee.Filter.eq('country_na','Benin')).set('ORIG_FID',0)

    # Prepare the cloud masked LANDSAT 7 median composite image
    benin_input = ee.ImageCollection("LANDSAT/LE07/C01/T1_SR").filterDate('2007-01-01', '2008-12-31')

    # Remove clouds, clip it to the outline of Benin (with buffer)
    benin_input = (benin_input.map(mask_clouds_landsat).median().clip(benin_shape.geometry().buffer(10000)))

    # Create NDVI band, rename RGB
    ndvi_img = benin_input.normalizedDifference(thermalBands).rename(['NDVI'])
    rgb_img = benin_input.select(opticalBands).rename(['R','G','B'])

    benin_input = rgb_img.addBands(ndvi_img)

    return benin_input.unmask(0) 


def get_labels_image(as_double:bool = True) -> ee.Image:
    """ Retrieves the labels image for Benin data
    
        - Uses voronoi asset 'projects/ls-test-3-24/assets/voronoi_villages'
        - Creates a mask where areas in villages are 1 and outisde are 0
        - Geometry created based on get_inputs_image return


    Args:
      as_double (bool): Whether to cast the image 
      to a double. This is done to match the 
      dtype of the input image

    Returns:
      labels image as ee.Image with 'target' band
    """

    # Import vornoi a ('ls-test-3-24/assets/voronoi_villages')
    # and convert to feature collection
    treated_voronoi = ee.FeatureCollection('projects/ls-test-3-24/assets/voronoi_villages')\
                                .filter(ee.Filter.eq('treated', 1))

    # Create a village mask based on the treated village raster
    villagemask = treated_voronoi.filter(ee.Filter.notNull(['VID']))\
                                .reduceToImage(properties=['VID'],reducer= ee.Reducer\
                                .first())\
                                .mask()

    # Create the target image using the village mask prepared earlier.
    inputs = get_inputs_image()
    l7Masked = inputs.updateMask(villagemask)

    # This part is hacky. How could we do it better?
    l7Unmasked = l7Masked.unmask(-9999)
    outside_circle = 'b("R") > -9000'
    target = l7Unmasked.expression(outside_circle).rename("target")

    if as_double: target = target.double()

    return target


def sample_points(seed: int = 0, points_per_class = 2) -> Iterable[tuple[float, float]]:
    """ Generates coordinate tuple of coordinates for sampling based on the labels image
        - Order of point generatiom: inside outside villages then in

    Args:
      seed (int): seed for stratifiedSample
      points_per_class (int): number of points per class in get_labels_image return

    Returns:
        Itterable of coordinates (tuple[float,float])
    """

    target = get_labels_image(as_double=False)
    benin = ee.FeatureCollection("USDOS/LSIB_SIMPLE/2017").filter(ee.Filter.eq('country_na','Benin')).set('ORIG_FID',0)

    points = target.stratifiedSample(
        points_per_class,
        scale=SAMPLE_SCALE,
        region=benin.geometry(),
        seed=seed,
        geometries=True,
    )

    for point in points.toList(points.size()).getInfo():
        yield point["geometry"]["coordinates"]





#XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
# The code below should not need to change between projects (how patches are retrieved)            #
#XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX


def get_inputs_patch(point: tuple, patch_size: int) -> np.ndarray:
    """Gets the patch of pixels for the inputs.

    Args:
        point: A (longitude, latitude) coordinate.
        patch_size: Size in pixels of the surrounding square patch.

    Returns: The pixel values of a patch as a NumPy array.
    """
    image = get_inputs_image()
    patch = get_patch(image, point, patch_size, PATCH_SCALE)


    return structured_to_unstructured(patch)

def get_labels_patch(point: tuple, patch_size: int) -> np.ndarray:
    """Gets the patch of pixels for the labels.

    Args:
        point: A (longitude, latitude) coordinate.
        patch_size: Size in pixels of the surrounding square patch.

    Returns:
        The pixel values of a patch as a NumPy array.
    """
    image = get_labels_image()
    patch = get_patch(image, point, patch_size, PATCH_SCALE)


    return structured_to_unstructured(patch)


@retry.Retry(deadline=10 * 60)  # Requires retries to work within 10 min
def get_patch(image: ee.Image, lonlat: tuple[float, float], patch_size: int, scale: int) -> np.ndarray:
    """Gets the patch of pixels based on parameters.

    Args:
        point: A (longitude, latitude) coordinate.
        patch_size: Size in pixels of the surrounding square patch.

    Returns:
        The pixel values of a patch as a NumPy array.
    """
    geometry = ee.Geometry.Point(lonlat)
    url = image.getDownloadURL({
        "region": geometry.buffer(scale * patch_size / 2, 1).bounds(1),
        "dimensions": [patch_size, patch_size],
        "format": "NPY",
    })

    # If we get "429: Too Many Requests" errors, it's safe to retry the request.
    # The Retry library only works with `google.api_core` exceptions.
    response = requests.get(url)
    if response.status_code == 429:
        raise exceptions.TooManyRequests(response.text)

    # Still raise any other exceptions to make sure we got valid data.
    response.raise_for_status()
    return np.load(io.BytesIO(response.content), allow_pickle=True)



